clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -1085,6 +1085,68 @@ def CmpOp : CIR_Op<"cmp", [Pure, SameTypeOperands]> {
   let hasVerifier = 0;
 }
 
+//===----------------------------------------------------------------------===//
+// CheckedArithOp
+//===----------------------------------------------------------------------===//
+
+def CAK_Add : I32EnumAttrCase<"Add", 1, "add">;
+def CAK_Sub : I32EnumAttrCase<"Sub", 2, "sub">;
+def CAK_Mul : I32EnumAttrCase<"Mul", 3, "mul">;
+
+def CheckedArithKind : I32EnumAttr<
+    "CheckedArithKind",
+    "checked arithmetic operation kind",
+    [CAK_Add, CAK_Sub, CAK_Mul]> {
+  let cppNamespace = "::mlir::cir";
+}
+
+def CheckedArithOp : CIR_Op<"checked_arith", [Pure, SameTypeOperands]> {
+  let summary = "Perform checked integral arithmetic";
+  let description = [{
+    `cir.checked_arith` performs arithmetic operations with overflow checking on
+    integral operands.
+
+    The `kind` argument specifies the kind of arithmetic operation to perform.
+    It can be either `add`, `sub`, or `mul`. The `lhs` and `rhs` arguments
+    specify the input operands of the arithmetic operation. The types of `lhs`
+    and `rhs` must be the same.
+
+    `cir.checked_arith` produces two SSA values. `result` is the result of the
+    arithmetic operation truncated to its specified type. `overflow` is a
+    boolean value indicating whether overflow happens during the operation.
+
+    The exact semantic of this operation is as follows:
+
+      - `lhs` and `rhs` are promoted to an imaginary integral type that has
+        infinite precision.
+      - The arithmetic operation is performed on the promoted operands.
+      - The infinite-precision result is truncated to the type of `result`. The
+        truncated result is assigned to `result`.
+      - If the truncated result is equal to the un-truncated result, `overflow`
+        is assigned to false. Otherwise, `overflow` is assigned to true.
```

Comment:
Great docs! Might be worth a quick note saying that these cover existing builtins.